### DISCUSSION TITLE
I switched the HC550 in the ZFS pool from 512e to 4Kn, and something strange happened.

### DISCUSSION THREAD

**Discussion by daiaji** [openzfs/zfs#15641] [General]
After converting a HC550 disk in the ZFS pool to 4Kn, I was surprised to find that I could directly bring that hard drive into an ONLINE state. My initial expectation was that I would need to repartition and rebuild the ZFS pool, but it seems that may not be necessary. Now, I would like to verify the integrity of the ZFS pool. Is it sufficient to use the `zpool scrub` command for this purpose?

From 512e to 4Kn, I created a new partition table and partitions, but did not format them. It seems that ZFS found the data that was not destroyed and then prevented the data from being rebuilt. 
ashift was already set to 12 when creating the pool. What should I do next? What do you think about switching from 512e to 4Kn?

**Comment by amotin** [comment]
ZFS unlike some other FS'es does not operate sectors internally, using byte offsets, aligned to ashift.  Since ashift selected covers both cases equally, nothing has changed for ZFS.  I am surprised you could reformat the drive without loosing the data, and recreated partition table fit into exactly the same offsets, but after that about ZFS I am mush less surprised.  Yes, scrub should be sufficient to verify that everything is still properly readable.

**Reply by daiaji** [reply]
Perhaps this is the benefit of 4K Alignment? Does this allow me to make ZFS work in the case of rebuilding partitioned tables?

I honestly don't know what happened, the impression seems to be that after the conversion to 4Kn, the partition table seemed to become very strange, then I rebuilt the GPT partition table and added a new partition.

Prior to this I basically followed Archwiki's guidelines to convert the logical sector from 512e to 4Kn. Contrary to what the guide says, Bing told me that this conversion did not damage my data.

https://wiki.archlinux.org/title/Advanced_Format#Advanced_Format_hard_disk_drives

**Reply by daiaji** [reply]
I switched the logical sector size of the second hard drive, and then the partition table became strange. Fdisk told me the hard drive was using a PMBR partition table. I then created a new partition table using GPT and added a partition. Finally, my hard drive was able to go ONLINE in the ZFS pool.

**Reply by rincebrain** [reply]
The guide is generally correct, switching sector sizes usually requires a low-level reformat which will eat your bits.

That said, partition tables are, for both GPT and MBR, in terms of logical sector size of the disk, so if you switch that, its reading of the partition table will produce impossible results (I know this from a number of USB enclosures that forge the sector size responses for the drive inside them, and surprises if you move the drives in/out of them.)
